### PR TITLE
Prevent null-ref and log exceptions form Serializer

### LINKED ISSUE #64: Access null object in Microsoft.AspNetCore.Antiforgery.Internal.DefaultAntiforgeryTokenSerializer.Deserialize
When I ran my app with debugger attached on Linux, I always hit fault on accessing invalid address. It turned out due to missing antiforgery cookie in the client request. So basically the test client faked requests and avoided using antiforgery token. The server code handling this is not checking whether this cookie exists or not before handling to deserializer.

```
Process 42821 stopped
* thread #15: tid = 42953, 0x00007fff7e0db40e, name = 'StarterMvc', stop reason = invalid address (fault address: 0x8)
    frame #0: 0x00007fff7e0db40e
-> 0x7fff7e0db40e: movl   0x8(%rbx), %r14d
   0x7fff7e0db412: movl   %r14d, %edi
   0x7fff7e0db415: callq  0x7fff7e0c5c90
   0x7fff7e0db41a: movl   %eax, %esi
(lldb) bt
* thread #15: tid = 42953, 0x00007fff7e0db40e, name = 'StarterMvc', stop reason = invalid address (fault address: 0x8)
  * frame #0: 0x00007fff7e0db40e
    frame #1: 0x00007fff7e0daeab
    frame #2: 0x00007fff7e0dad84
    frame #3: 0x00007fff7e0dab6e
    frame #4: 0x00007fff7e0da86e
    frame #5: 0x00007fff7e0da383
  …
    frame #90: 0x00007fff7c9d3215
    frame #91: 0x00007ffff6420fa7 libcoreclr.so`CallDescrWorkerInternal + 124
    frame #92: 0x00007ffff633c09e libcoreclr.so`MethodDescCallSite::CallTargetWorker(unsigned long const*) + 862
    frame #93: 0x00007ffff6538a0a libcoreclr.so`QueueUserWorkItemManagedCallback(void*) + 186
    frame #94: 0x00007ffff630ff8e libcoreclr.so`ManagedThreadBase_DispatchOuter(ManagedThreadCallState*) + 430
    frame #95: 0x00007ffff6310736 libcoreclr.so`ManagedThreadBase::ThreadPool(ADID, void (*)(void*), void*) + 70
    frame #96: 0x00007ffff651a974 libcoreclr.so`ManagedPerAppDomainTPCount::DispatchWorkItem(bool*, bool*) + 292
    frame #97: 0x00007ffff632e8ec libcoreclr.so`ThreadpoolMgr::WorkerThreadStart(void*) + 1228
    frame #98: 0x00007ffff67d96c9 libcoreclr.so`CorUnix::CPalThread::ThreadEntry(pvParam=0x0000000000832cf0) + 297 at thread.cpp:1641
    frame #99: 0x00007ffff79c0182 libpthread.so.0`start_thread + 194
    frame #100: 0x00007ffff6ecd47d libc.so.6`clone + 109
```

The missing stacks are managed stack:

```
(lldb) clrstack
OS Thread Id: 0xa7c9 (15)
        Child SP               IP Call Site
00007FFF5FFFA6E0 00007FFF7E0DB40E Microsoft.AspNetCore.Antiforgery.Internal.DefaultAntiforgeryTokenSerializer.Deserialize(System.String)
00007FFF5FFFA730 00007FFF7E0DAEAB Microsoft.AspNetCore.Antiforgery.Internal.DefaultAntiforgery.GetCookieTokenDoesNotThrow(Microsoft.AspNetCore.Http.HttpContext)
00007FFF5FFFA760 00007FFF7E0DAD84 Microsoft.AspNetCore.Antiforgery.Internal.DefaultAntiforgery.GetCookieTokens(Microsoft.AspNetCore.Http.HttpContext)
00007FFF5FFFA790 00007FFF7E0DAB6E Microsoft.AspNetCore.Antiforgery.Internal.DefaultAntiforgery.GetTokensInternal(Microsoft.AspNetCore.Http.HttpContext)
00007FFF5FFFA7C0 00007FFF7E0DA86E Microsoft.AspNetCore.Antiforgery.Internal.DefaultAntiforgery.GetAndStoreTokens(Microsoft.AspNetCore.Http.HttpContext)
00007FFF5FFFA7F0 00007FFF7E0DA383 Microsoft.AspNetCore.Mvc.ViewFeatures.AntiforgeryExtensions.GetHtml(Microsoft.AspNetCore.Antiforgery.IAntiforgery, 
```

In the code:

```c#
        private AntiforgeryToken GetCookieTokenDoesNotThrow(HttpContext httpContext)
        {
            try
            {
                var serializedToken = _tokenStore.GetCookieToken(httpContext);
                var token = _tokenSerializer.Deserialize(serializedToken);

                return token;
            }
            catch
            {
                // ignore failures since we'll just generate a new token
                return null;
            }
…
        public AntiforgeryToken Deserialize(string serializedToken)
        {
            var serializationContext = _pool.Get();

            Exception innerException = null;
            try
            {
                var count = serializedToken.Length;
```

Although this exception would be captured and ignored, it incurs extra cost, and could be easily avoided by checking if the searializedToken is null.

### PR DESCRIPTION
**Pull request by ryanbrandenburg**
Fixes #64.

@rynowak is this what you had in mind?

### CODE REVIEW

File: src/Microsoft.AspNetCore.Antiforgery/Internal/DefaultAntiforgery.cs [hunk 1]
```diff
@@ -329,15 +329,20 @@ private AntiforgeryToken GetCookieTokenDoesNotThrow(HttpContext httpContext)
             try
             {
                 var serializedToken = _tokenStore.GetCookieToken(httpContext);
-                var token = _tokenSerializer.Deserialize(serializedToken);
 
-                return token;
+                if (!string.IsNullOrEmpty(serializedToken))
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Antiforgery/Internal/DefaultAntiforgery.cs]: When would `serializedToken` be empty?

**Comment by rynowak** [comment]
:watch: - yes this is the right idea. Needs a test

**Comment by ryanbrandenburg** [comment]
:up::date:

File: test/Microsoft.AspNetCore.Antiforgery.Test/Internal/DefaultAntiforgeryTest.cs [hunk 2]
```diff
@@ -928,6 +928,52 @@ public void SetCookieTokenAndHeader_DoesNotStoreTwice()
...
+        public void SetCookieTokenAndHeader_NullCookieToken()
+        {
+            //Console.WriteLine($"PROCESS ID: {System.Diagnostics.Process.GetCurrentProcess().Id}");
+            //Console.Out.Flush();
+            //while (!System.Diagnostics.Debugger.IsAttached)
+            //{
+            //    System.Threading.Thread.Sleep(1000);
+            //}
+
+            //System.Diagnostics.Debugger.Break();
```
Comment (rynowak) [inline test/Microsoft.AspNetCore.Antiforgery.Test/Internal/DefaultAntiforgeryTest.cs]: :crying_cat_face:

**Comment by rynowak** [comment]
:shipit:

**Pull request merged**